math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
+        private final int fence; // one past the last column to process
+
+        SparseMatrixSpliterator(SparseMatrix matrix, int col, int fence) {
+            this.m = matrix;
+            this.col = col;
+            this.index = m.colIndex[col];
+            this.fence = fence;
+        }
+
+        public void forEachRemaining(Consumer<? super SparseMatrixEntry> action) {
+            for (; col < fence; col++) {
+                for (; index < m.colIndex[col + 1]; index++) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                }
+            }
+        }
+
+        public boolean tryAdvance(Consumer<? super SparseMatrixEntry> action) {
+            if (col < fence) {
+                while (col < fence && index >= m.colIndex[col + 1]) {
+                    col++;
+                }
+                if (col < fence) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                    index++;
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                // cannot advance
+                return false;
+            }
+        }
+
+        public Spliterator<SparseMatrixEntry> trySplit() {
+            int lo = col; // divide range in half, assuming roughly constant size of all columns
+            int mid = ((lo + fence) >>> 1) & ~1; // force midpoint to be even
+            if (lo < mid) { // split out left half
+                col = mid; // reset this Spliterator's origin
+                return new SparseMatrixSpliterator(m, lo, mid);
+            } else {
+                // too small to split
+                return null;
+            }
+        }
+
+        public long estimateSize() {
+            return (long) (m.colIndex[fence] - m.colIndex[col]);
+        }
+
+        public int characteristics() {
+            return ORDERED | SIZED | IMMUTABLE | SUBSIZED;
+        }
+    }
+
+    /**
+     * Encapsulates important information about an entry in a matrix for use
+     * in streaming, including a string that leads back to the original cell
+     * so that in-place updates are possible.
+     */
+    public static class SparseMatrixEntry {
+        // these fields are exposed for direct access to simplify in-lining by the JVM
+        public final int row;
+        public final int col;
+        public double x;
+
+        // these are hidden due to internal dependency
+        private final int index;
+        private double[] values;
+
+
+        @SuppressWarnings("WeakerAccess")
+        public SparseMatrixEntry(int row, int col, double x, int index, double[] values) {
+            this.row = row;
+            this.col = col;
+            this.x = x;
+            this.index = index;
+            this.values = values;
+        }
+
+        public void set(double value) {
```

Comment:
This breaks our contract that `SparseMatrix` is immutable.

Reply:
The contract was already broken. And was undocumented.

Reply:
In v2, we enhance the Matrix interface. Only `DenseMatrix` is mutable.

Reply:
You have just made pronouncements so far. Without technical justification, those are just opinions.

What is the technical rationale? I have provided context from the many other sparse matrix packages I have worked on that mutation on sparse matrices is only a problem due to brittleness of typical csr or rsr representations. This idiom prevents that problem.